chain-storage/src/lib.rs

```diff
@@ -1,604 +1,424 @@
-        match self
-            .inner
-            .prepare_cached("select hash from Tags where name = ?")
-            .map_err(|err| Error::BackendError(Box::new(err)))?
-            .query_row(&[&tag_name], |row| Ok(blob_to_hash(row.get(0)?)))
-        {
-            Ok(s) => Ok(Some(s)),
-            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
-            Err(err) => Err(Error::BackendError(Box::new(err))),
+        while let Some(current_tip) = &maybe_current_tip {
+            let result = (&blocks, &info, &height_to_block_ids, &tips).transaction(
+                |(blocks, info, height_to_block_ids, tips)| {
+                    let current_tip: &[u8] = current_tip;
+
+                    let block_info_bin = info.remove(current_tip)?.unwrap();
+                    let mut block_info_reader: &[u8] = &block_info_bin;
+                    let block_info = BlockInfo::deserialize(&mut block_info_reader);
+
+                    blocks.remove(current_tip)?;
+
+                    let mut height_index = block_info.chain_length().to_le_bytes().to_vec();
+                    height_index.extend_from_slice(block_info.id());
+                    height_to_block_ids.remove(height_index)?;
+
+                    tips.remove(current_tip)?;
+
+                    if block_info.parent_id() == vec![0; block_info.parent_id().len()].as_slice() {
+                        return Ok(Ok(None));
+                    }
+
+                    let parent_block_info_bin = info.get(block_info.parent_id())?.unwrap();
+                    let mut parent_block_info_reader: &[u8] = &parent_block_info_bin;
+                    let mut parent_block_info =
+                        BlockInfo::deserialize(&mut parent_block_info_reader);
+                    parent_block_info.remove_ref();
+                    info.insert(parent_block_info.id(), parent_block_info.serialize())?;
+
+                    let maybe_next_tip = if parent_block_info.ref_count == 0 {
+                        // if the block is inside another branch it cannot be a tip
+                        tips.insert(block_info.parent_id(), &[])?;
+                        Some(block_info.parent_id().to_vec())
+                    } else {
+                        None
+                    };
+
+                    Ok(Ok(maybe_next_tip))
+                },
+            );
+
+            // Transactional impl for (&Tree, &Tree) implies the use of () as
+            // the type for user-defined errors, so we have a workaround for
+            // that.
+            match result {
+                Ok(Ok(maybe_next_tip)) => {
+                    maybe_current_tip = maybe_next_tip;
+                }
+                Ok(Err(err)) => return Err(err),
+                Err(TransactionError::Storage(err)) => {
+                    return Err(Error::BackendError(Box::new(err)))
+                }
+                Err(TransactionError::Abort(())) => unreachable!(),
```

Comment:
can't we do the pruning in one transaction instead?

Reply:
Technically not a big problem. But if a branch is too big this transaction will lock the database for a relatively long time (need to measure that though).